src/Microsoft.AspNetCore.Blazor.Razor.Extensions/BlazorRuntimeNodeWriter.cs

```diff
@@ -373,6 +363,173 @@ private void WriteElementAttributeToken(CodeRenderingContext context, HtmlTagTok
+            _scopeStack.IncrementCurrentScopeChildCount(context);
+
+            // builder.OpenComponent<TComponent>(42);
+            context.CodeWriter.Write(_scopeStack.BuilderVarName);
+            context.CodeWriter.Write(".");
+            context.CodeWriter.Write(BlazorApi.RenderTreeBuilder.CloseComponent);
+            context.CodeWriter.Write("();");
+            context.CodeWriter.WriteLine();
+        }
+
+        public override void WriteComponentBody(CodeRenderingContext context, ComponentBodyExtensionNode node)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if (node == null)
+            {
+                throw new ArgumentNullException(nameof(node));
+            }
+
+            _scopeStack.OpenScope(node.TagName, isComponent: true);
+            context.RenderChildren(node);
+            _scopeStack.CloseScope(context, node.TagName, isComponent: true, source: node.Source);
+        }
+
+        public override void WriteComponentAttribute(CodeRenderingContext context, ComponentAttributeExtensionNode node)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if (node == null)
+            {
+                throw new ArgumentNullException(nameof(node));
+            }
+
+            // builder.OpenComponent<TComponent>(42);
+            context.CodeWriter.Write(_scopeStack.BuilderVarName);
+            context.CodeWriter.Write(".");
+            context.CodeWriter.Write(BlazorApi.RenderTreeBuilder.AddAttribute);
+            context.CodeWriter.Write("(");
+            context.CodeWriter.Write((_sourceSequence++).ToString());
+            context.CodeWriter.Write(", ");
+            context.CodeWriter.WriteStringLiteral(node.AttributeName);
+            context.CodeWriter.Write(", ");
+
+            if (node.AttributeStructure == AttributeStructure.Minimized)
+            {
+                // Minimized attributes always map to 'true'
+                context.CodeWriter.Write("true");
+            }
+            else if (
+                node.Children.Count != 1 ||
+                node.Children[0] is HtmlContentIntermediateNode htmlNode && htmlNode.Children.Count != 1 ||
+                node.Children[0] is CSharpExpressionIntermediateNode cSharpNode && cSharpNode.Children.Count != 1)
+            {
+                // We don't expect this to happen, we just want to know if it can.
+                throw new InvalidOperationException("Attribute nodes should either be minimized or a single content node.");
+            }
+            else if (node.BoundAttribute.IsUIEventHandlerProperty())
+            {
+                // This is a UIEventHandler property. We do some special code generation for this
+                // case so that it's easier to write for common cases.
```

Comment:
Note, this isn't up to date based on our conversation this morning.

Reply:
A further update would generalise this to arbitrary delegate types (not just `UIEventHandler`), would it?

Reply:
yes, that's the idea